src/front/wgsl/ast.rs

```diff
@@ -0,0 +1,344 @@
+
+#[derive(Debug)]
+pub struct GlobalDecl<'a> {
+    pub kind: GlobalDeclKind<'a>,
+    pub dependencies: FastHashSet<Dependency<'a>>,
+}
+
+#[derive(Debug)]
+pub enum GlobalDeclKind<'a> {
+    Fn(Function<'a>),
+    Var(GlobalVariable<'a>),
+    Const(Const<'a>),
+    Struct(Struct<'a>),
+    Type(TypeAlias<'a>),
+}
+
+#[derive(Debug)]
+pub struct FunctionArgument<'a> {
+    pub name: Ident<'a>,
+    pub ty: Type<'a>,
+    pub binding: Option<crate::Binding>,
+    pub handle: Handle<Local>,
+}
+
+#[derive(Debug)]
+pub struct FunctionResult<'a> {
+    pub ty: Type<'a>,
+    pub binding: Option<crate::Binding>,
+}
+
+#[derive(Debug)]
+pub struct EntryPoint {
+    pub stage: crate::ShaderStage,
+    pub early_depth_test: Option<crate::EarlyDepthTest>,
+    pub workgroup_size: [u32; 3],
+}
+
+#[derive(Debug)]
+pub struct Function<'a> {
+    pub entry_point: Option<EntryPoint>,
+    pub name: Ident<'a>,
+    pub arguments: Vec<FunctionArgument<'a>>,
+    pub result: Option<FunctionResult<'a>>,
+    pub expressions: Arena<Expression<'a>>,
+    pub locals: Arena<Local>,
+    pub body: Block<'a>,
+}
+
+#[derive(Debug)]
+pub struct GlobalVariable<'a> {
+    pub name: Ident<'a>,
+    pub space: crate::AddressSpace,
+    pub binding: Option<crate::ResourceBinding>,
+    pub ty: Type<'a>,
+    pub init: Option<Handle<Expression<'a>>>,
+}
+
+#[derive(Debug)]
+pub struct StructMember<'a> {
+    pub name: Ident<'a>,
+    pub ty: Type<'a>,
+    pub binding: Option<crate::Binding>,
+    pub align: Option<(u32, Span)>,
+    pub size: Option<(u32, Span)>,
+}
+
+#[derive(Debug)]
+pub struct Struct<'a> {
+    pub name: Ident<'a>,
+    pub members: Vec<StructMember<'a>>,
+}
+
+#[derive(Debug)]
+pub struct TypeAlias<'a> {
+    pub name: Ident<'a>,
+    pub ty: Type<'a>,
+}
+
+#[derive(Debug)]
+pub struct Const<'a> {
+    pub name: Ident<'a>,
+    pub ty: Option<Type<'a>>,
+    pub init: Handle<Expression<'a>>,
+}
+
+#[derive(Debug)]
+pub struct Type<'a> {
+    pub kind: TypeKind<'a>,
+    pub span: Span,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum ArraySize<'a> {
+    Constant(Handle<Expression<'a>>),
+    Dynamic,
+}
+
+#[derive(Debug)]
+pub enum TypeKind<'a> {
+    Scalar {
+        kind: crate::ScalarKind,
+        width: crate::Bytes,
+    },
+    Vector {
+        size: crate::VectorSize,
+        kind: crate::ScalarKind,
+        width: crate::Bytes,
+    },
+    Matrix {
+        columns: crate::VectorSize,
+        rows: crate::VectorSize,
+        width: crate::Bytes,
+    },
+    Atomic {
+        kind: crate::ScalarKind,
+        width: crate::Bytes,
+    },
+    Pointer {
+        base: Box<Type<'a>>,
+        space: crate::AddressSpace,
+    },
+    Array {
+        base: Box<Type<'a>>,
+        size: ArraySize<'a>,
+    },
+    Image {
+        dim: crate::ImageDimension,
+        arrayed: bool,
+        class: crate::ImageClass,
+    },
+    Sampler {
+        comparison: bool,
+    },
+    BindingArray {
+        base: Box<Type<'a>>,
```

Comment:
I think we can replace the 3 boxes in `TypeKind` with `Handle`s (and an additional `Arena`).

Reply:
I'm not sure this is worth it either, as type nesting is rare, and is also usually quite shallow.

Reply:
I agree that most types won't be nested but we are dealing with user input here, so we need to expect the worst scenario where we'll have a bunch of `Box`es spread out in memory.